examples/epocher/src/application/actor.rs

```diff
@@ -0,0 +1,260 @@
+                        },
+                        valid = self.verify_block(context, payload) => {
+                            let _ = response.send(valid);
+                        }
+                    }
+                }
+                Message::Report { block, response } => {
+                    let height = block.height;
+                    info!(height, "finalized-delivered-to-app");
+
+                    // Return early if not the last block in the epoch.
+                    let Some(epoch) = epoch::is_last_block_in_epoch(height) else {
+                        continue;
+                    };
+
+                    // Respond to any responders for this block height.
+                    if let Some(responders) = self.responders.remove(&height) {
+                        for responder in responders {
+                            let _ = responder.send(block.commitment());
+                        }
+                    }
+
+                    // Get the finalization for this block.
+                    let cert = match epoch {
+                        0 => {
+                            let Some(f0) = self.get_finalization(height).await else {
+                                continue;
+                            };
+                            EpochCert::Single(f0)
+                        }
+                        _ => {
+                            let Some(f2) = self.get_finalization(height).await else {
+                                continue;
+                            };
+                            let previous_height = epoch::get_last_height(epoch - 1);
+                            let Some(f1) = self.get_finalization(previous_height).await else {
+                                continue;
+                            };
+                            EpochCert::Double(f1, f2)
+                        }
+                    };
+
+                    orchestrator.report(cert).await;
+
+                    let _ = response.send(());
+                }
+            }
+        }
+    }
+
+    /// Gets the finalization for a given height from marshal.
+    async fn get_finalization(
+        &mut self,
+        height: u64,
+    ) -> Option<Finalization<MinSig, Sha256Digest>> {
+        let Ok(Some((finalization, _block))) = self.marshal.get_finalization(height).await.await
+        else {
+            warn!("finalization not found for height {}", height);
+            return None;
+        };
+        Some(finalization)
+    }
+
+    /// Proposes a new block.
+    async fn propose_block(&mut self, context: Context<Sha256Digest>) -> Sha256Digest {
+        let round = context.round;
+        let epoch = round.epoch();
+        let (parent_view, parent_commitment) = context.parent;
+        let parent_round = Round::new(epoch, parent_view);
+
+        // Get the parent block from marshal.
+        let parent = self.subscribe_block(parent_round, parent_commitment).await;
+
+        // Case: Reproposal.
```

Comment:
As discussed, I wonder if there is a way to hide this from the application in `marshal` (I feel like an application could only "mess it up" by providing something wrong that builds on this)?